### ISSUE TITLE
Problem uploading to blobstore

### ISSUE THREAD

**Issue by shaneh2** [Azure/azure-storage-azcopy#191]
### Which version of the AzCopy was used? 
AzCopy 10.0.7-Preview

### Which platform are you using? (ex: Windows, Mac, Linux)
Windows 10

### What command did you run?
./azcopy cp SOURCE_DIRECTORY REMOTE_BLOBSTORE_SAS_STRING  --overwrite

### What problem was encountered?
The upload worked fine for a while, but then produced this error message:
 It worked fine for a while, but then I got this error message:

```
8 Done, 917 Failed, 55 Pending, 0 Skipped, 980 Total , 2-sec Throughput (Mb/s): 79.475panic: runtime error: slice bounds
 out of range

goroutine 2835 [running]:
github.com/Azure/azure-storage-azcopy/common.(*singleChunkReader).doRead(0xc0429b3dd0, 0xc07fe09212, 0xdee, 0xdee, 0x500
601, 0xc07e4685c0, 0x1, 0x400000000000000)
        /go/src/github.com/Azure/azure-storage-azcopy/common/singleChunkReader.go:254 +0x183
github.com/Azure/azure-storage-azcopy/common.(*singleChunkReader).Read(0xc0429b3dd0, 0xc07fe09212, 0xdee, 0xdee, 0x0, 0x
c08005da48, 0x501480)
        /go/src/github.com/Azure/azure-storage-azcopy/common/singleChunkReader.go:235 +0x55
github.com/Azure/azure-storage-azcopy/ste.(*liteBodyPacer).Read(0xc07cbfa6e0, 0xc07fe09212, 0xdee, 0xdee, 0xc08005db88,
0x65fdc2, 0xbcd1a0)
        /go/src/github.com/Azure/azure-storage-azcopy/ste/pacer-lite.go:50 +0x58
github.com/Azure/azure-storage-azcopy/vendor/github.com/Azure/azure-pipeline-go/pipeline.(*retryableRequestBody).Read(0x
c0469c5bc0, 0xc07fe09212, 0xdee, 0xdee, 0x65fdc2, 0x1, 0x1)
        /go/src/github.com/Azure/azure-storage-azcopy/vendor/github.com/Azure/azure-pipeline-go/pipeline/request.go:129
+0x58
net/http.transferBodyReader.Read(0xc07f131b80, 0xc07fe09212, 0xdee, 0xdee, 0x40d504, 0xc07cd38000, 0xa76100)
        /usr/local/go/src/net/http/transfer.go:60 +0x5d
io.(*LimitedReader).Read(0xc07cbfa9c0, 0xc07fe09212, 0xdee, 0xdee, 0xc042af4100, 0xc042273590, 0x4eb104)
        /usr/local/go/src/io/io.go:446 +0x6a
bufio.(*Writer).ReadFrom(0xc046a58a40, 0xbc6da0, 0xc07cbfa9c0, 0x1a4cbf8, 0xc046a58a40, 0x1)
        /usr/local/go/src/bufio/bufio.go:703 +0xd4
io.copyBuffer(0xbc5720, 0xc046a58a40, 0xbc6da0, 0xc07cbfa9c0, 0x0, 0x0, 0x0, 0xa9df20, 0xc042a8a001, 0xc07cbfa9c0)
        /usr/local/go/src/io/io.go:386 +0x321
io.Copy(0xbc5720, 0xc046a58a40, 0xbc6da0, 0xc07cbfa9c0, 0xc07fe09210, 0xb3fa50, 0x2)
        /usr/local/go/src/io/io.go:362 +0x61
net/http.(*transferWriter).WriteBody(0xc07f131b80, 0xbc5720, 0xc046a58a40, 0x2, 0x2)
        /usr/local/go/src/net/http/transfer.go:337 +0x601
net/http.(*Request).write(0xc042822f00, 0xbc5720, 0xc046a58a40, 0x0, 0xc042273590, 0x0, 0x0, 0x0)
        /usr/local/go/src/net/http/request.go:622 +0x6b9
net/http.(*persistConn).writeLoop(0xc07f0e10e0)
        /usr/local/go/src/net/http/transport.go:1825 +0x1f1
created by net/http.(*Transport).dialConn
        /usr/local/go/src/net/http/transport.go:1238 +0x986
```

### How can we reproduce the problem in the simplest way?

I am not sure.

### Have you found a mitigation/solution?

No.

**Comment by JohnRusk** [comment]
Ah, I see, you had thousands of files, as shown by the first line you posted.  Let me see what I can find - the call stack is fairly specific, so might be able to track it down with no further info.

**Comment by JohnRusk** [comment]
Hi again @shaneh2. I've spent the last few hours on this. The callstack shows something that "shouldn't" happen, so I'm particularly eager to find out what's going on.  So far, I've been unable to reproduce it or to deduce the likely cause. I notice that your screendump shows that a very large number of files failed. I wonder if those other failures in the log might shed some light on what's happening. Would you be able to zip up the log file, and email it to me?  AzCopy automatically scrubs all SAS tokens out of the URLs before it logs them, so you don't need to worry about those being exposed.  (It does not remove the filenames however; those will remain in the log.) 

If you can email me the log, that could be very helpful.

**Comment by shaneh2** [comment]
@JohnRusk Thank-you for taking the time to help me with this.. I have sent you an email with the logs you asked for. Please let me know if you need anything else from me which could possibly help.

**Comment by JohnRusk** [comment]
Thanks @shaneh2 .  While I have not been able to determine _exactly_ what caused your error, I have a theory and have put in place extra logging to prove or disprove that theory if/when it happens again. That logging will be in a future release (probably the next one that we do, but that's not 100% confirmed yet).  In the meantime, if you're able to reply to my offline email, with that additional log file, that could be useful.

**Comment by JohnRusk** [comment]
We are adding the extra diagnostics.  In the absence of any additional log file, that's all we can do for now.  Will leave this issue open for a while, in the hope that the diagnostics catch the issue "in the wild" and allow is to report the exact root cause here.

**Comment by JohnRusk** [comment]
We believe we've found the root cause of the error. Namely, when the context is cancelled (due to some earlier error) the reader object can get closed immediately - even if a read is still in progress.  So it wasn't a "real" error (in that the real error had happened earlier and was probably something like exhausting the retry count after network failures).  But it was poor and confusing handing _following_ the real error.  So we'll fix it.

**Comment by JohnRusk** [comment]
This issue is fixed in the latest release 10.0.9
